source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -260,8 +269,451 @@ static int s_aws_mqtt5_adapter_perform_safe_callback(
+
+    struct aws_mqtt_adapter_disconnect_task *task =
+        s_aws_mqtt_adapter_disconnect_task_new(adapter->allocator, adapter, on_disconnect, on_disconnect_user_data);
+    if (task == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_MQTT_CLIENT, "id=%p: failed to create adapter disconnect task", (void *)adapter);
+        return AWS_OP_ERR;
+    }
+
+    aws_event_loop_schedule_task_now(adapter->loop, &task->task);
+
+    return AWS_OP_SUCCESS;
+}
+
+struct aws_mqtt_adapter_connect_task {
+    struct aws_task task;
+    struct aws_allocator *allocator;
+    struct aws_mqtt_client_connection_5_impl *adapter;
+
+    struct aws_byte_buf host_name;
+    uint16_t port;
+    struct aws_socket_options socket_options;
+    struct aws_tls_connection_options *tls_options_ptr;
+    struct aws_tls_connection_options tls_options;
+
+    struct aws_byte_buf client_id;
+    uint16_t keep_alive_time_secs;
+    uint32_t ping_timeout_ms;
+    uint32_t protocol_operation_timeout_ms;
+    aws_mqtt_client_on_connection_complete_fn *on_connection_complete;
+    void *on_connection_complete_user_data;
+    bool clean_session;
+};
+
+static void s_aws_mqtt_adapter_connect_task_destroy(struct aws_mqtt_adapter_connect_task *task) {
+    if (task == NULL) {
+        return;
+    }
+
+    aws_byte_buf_clean_up(&task->host_name);
+    aws_byte_buf_clean_up(&task->client_id);
+
+    if (task->tls_options_ptr) {
+        aws_tls_connection_options_clean_up(task->tls_options_ptr);
+    }
+
+    aws_mem_release(task->allocator, task);
+}
+
+static void s_adapter_connect_task_fn(struct aws_task *task, void *arg, enum aws_task_status status);
+
+static struct aws_mqtt_adapter_connect_task *s_aws_mqtt_adapter_connect_task_new(
+    struct aws_allocator *allocator,
+    struct aws_mqtt_client_connection_5_impl *adapter,
+    const struct aws_mqtt_connection_options *connection_options) {
+
+    struct aws_mqtt_adapter_connect_task *connect_task =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_adapter_connect_task));
+
+    aws_task_init(&connect_task->task, s_adapter_connect_task_fn, (void *)connect_task, "AdapterConnectTask");
+    connect_task->allocator = adapter->allocator;
+    connect_task->adapter = (struct aws_mqtt_client_connection_5_impl *)aws_ref_count_acquire(&adapter->internal_refs);
+
+    aws_byte_buf_init_copy_from_cursor(&connect_task->host_name, allocator, connection_options->host_name);
+    connect_task->port = connection_options->port;
+    connect_task->socket_options = *connection_options->socket_options;
+    if (connection_options->tls_options) {
+        aws_tls_connection_options_copy(&connect_task->tls_options, connection_options->tls_options);
+        connect_task->tls_options_ptr = &connect_task->tls_options;
+    }
+
+    aws_byte_buf_init_copy_from_cursor(&connect_task->client_id, allocator, connection_options->client_id);
+
+    connect_task->keep_alive_time_secs = connection_options->keep_alive_time_secs;
```

Comment:
Tangent: If we aren't already elsewhere, should we force/change/check that the ping_timeout_ms is less than the keep_alive_time_secs?

Reply:
Sure, will add